README.md
# TinyDraw

[![Build](https://github.com/tinygo-org/tinydraw/actions/workflows/build.yml/badge.svg?branch=dev)](https://github.com/tinygo-org/tinydraw/actions/workflows/build.yml)

TinyDraw is a package for drawing primitives on [TinyGo](https://tinygo.org/) displays. It is heavily based on the [Adafruit GFX library](https://github.com/adafruit/Adafruit-GFX-Library).

![example](https://raw.githubusercontent.com/tinygo-org/tinydraw/master/example.png)

This package is experimental and may change in the future. It has not been optimized for speed or memory usage (yet).
## License

[This project is licensed](./LICENSE) under the BSD 3-clause license, just like the [Go project](https://golang.org/LICENSE) itself.


CHANGELOG.md
0.4.0
---

- **build**
    - switch to ghcr.io for docker container
    - switch to github actions
- **docs**
    - switch badge to GH actions
    - update README.md
- **test**
    - correct Go formatting for test file
- **modules**
    - update to latest tinygo drivers
- **misc**
    - add .gitignore


tinydraw_test.go
package tinydraw

import (
	"fmt"
	"image"
	"image/color"
	"image/png"
	"math"
	"os"
	"path/filepath"
	"testing"
	"time"
)

// Set to true to generate missing reference images for new tests
const generateMissingReferenceImages = false

type testDisplay struct {
	*image.RGBA
}

func (d *testDisplay) Size() (x, y int16) {
	return int16(d.Rect.Dx()), int16(d.Rect.Dy())
}

func (d *testDisplay) SetPixel(x, y int16, c color.RGBA) {
	d.Set(int(x), int(y), c)
}

func newTestDisplay() testDisplay {
	i := image.NewRGBA(image.Rect(0, 0, 320, 240))
	return testDisplay{
		RGBA: i,
	}
}

func compareWithReference(name string, d testDisplay) error {

	referenceFilePath := filepath.Join("testdata", name+".png")

	referenceFile, err := os.Open(referenceFilePath)
	switch err.(type) {
	case *os.PathError:
		if generateMissingReferenceImages {
			saveImgFile, err := os.Create(referenceFilePath)
			if err != nil {
				return err
			}

			err = png.Encode(saveImgFile, d.RGBA)
			return err
		} else {
			return fmt.Errorf("Missing reference image %v. Set generateMissingReferenceImages in tinydraw_test.go to true to autogenerate.", referenceFilePath)
		}
	default:
		if err != nil {
			return err
		}
	}

	referenceImage, err := png.Decode(referenceFile)
	if err != nil {
		return err
	}

	sx, sy := d.Size()

	differenceDetected := false

	for y := 0; y < int(sy); y++ {
		for x := 0; x < int(sx); x++ {
			refPixel := referenceImage.At(x, y)
			testPixel := d.RGBA.At(x, y)

			refR, refG, refB, refA := refPixel.RGBA()
			testR, testG, testB, testA := testPixel.RGBA()

			if refR != testR || refG != testG || refB != testB || refA != testA {
				differenceDetected = true
				d.RGBA.Set(x, y, color.RGBA{255, 0, 0, 255}) //Mark the pixel as bad for debugging
			}
		}
	}

	if differenceDetected {
		diffFilePath := filepath.Join("testdata", "DIFF_"+name+".png")
		saveImgFile, err := os.Create(diffFilePath)
		if err != nil {
			return err
		}

		err = png.Encode(saveImgFile, d.RGBA)
		return fmt.Errorf("Image contains differences see %v" + diffFilePath)
	}

	return nil
}

// Display sends the buffer (if any) to the screen.
func (d *testDisplay) Display() error {
	return nil
}

func TestFilledTriangleClockMinuteHand(t *testing.T) {

	black := color.RGBA{0, 0, 0, 255}
	unfilledDisplay := newTestDisplay()
	filledDisplay := newTestDisplay()

	x := int16(160)
	y := int16(120)

	const (
		hourRadius   = 55
		minuteRadius = 95
		hourWidth    = 6
		minuteWidth  = 5
	)

	testTime, _ := time.Parse("15:04:05", "10:02:25")
	for i := 0; i < 60; i++ {
		// Draw the clock hands.
		minuteAngle := float64(testTime.Minute()) / 60 * 2 * math.Pi
		mx, my := math.Sincos(minuteAngle)

		x0 := x - int16(minuteWidth*my)
		y0 := y - int16(minuteWidth*mx)
		x1 := x + int16(minuteWidth*my)
		y1 := y + int16(minuteWidth*mx)
		x2 := x + int16(minuteRadius*mx)
		y2 := y - int16(minuteRadius*my)

		FilledTriangle(

			&filledDisplay,
			x0, y0,
			x1, y1,
			x2, y2,
			black)

		Triangle(
			&unfilledDisplay,
			x0, y0,
			x1, y1,
			x2, y2,
			black)

		testTime = testTime.Add(1 * time.Minute)
	}

	err := compareWithReference("TestFilledTriangleClockMinuteHand", filledDisplay)
	if err != nil {
		t.Errorf("Fail %v", err)
	}

	err = compareWithReference("TestTriangleClockMinuteHand", unfilledDisplay)
	if err != nil {
		t.Errorf("Fail %v", err)
	}
}

type triangleTest struct {
	name string
	x0   int16
	y0   int16
	x1   int16
	y1   int16
	x2   int16
	y2   int16
}

func TestTriangles(t *testing.T) {

	tests := []triangleTest{
		{"TopAndBottom1", 160, 0, 80, 240, 213, 120}, //Basic triangle with top and bottom half
		{"TopAndBottom2", 160, 0, 213, 240, 80, 120}, //Basic triangle with top and bottom half
		{"NoTop", 80, 50, 160, 50, 120, 200},         //No top
		{"NoBottom", 80, 200, 160, 200, 120, 50},     //No Bottom
		{"HorizLine", 80, 200, 160, 200, 213, 200},   //Horizontal line
		{"VertLine", 80, 80, 80, 120, 80, 200},       //Vertical line

		{"ClockHand1", 160, 115, 160, 125, 255, 120}, //Previously rendered too Wide
		{"ClockHand2", 160, 116, 160, 124, 254, 129}, //Previusly fill extendended to the left outside of triangle
		{"ClockHand3", 165, 120, 155, 120, 160, 215}, //Too Wide
	}

	colors := []color.RGBA{
		{255, 0, 0, 255},
		{0, 255, 0, 255},
		{0, 0, 255, 255},
		{255, 255, 0, 255},
		{0, 255, 255, 255},
		{0, 0, 0, 255},
	}

	for _, test := range tests {

		unfilledDisplay := newTestDisplay()

		//Test all permutations of coordinates
		Triangle(&unfilledDisplay, test.x0, test.y0, test.x1, test.y1, test.x2, test.y2, colors[0])
		Triangle(&unfilledDisplay, test.x0, test.y0, test.x2, test.y2, test.x1, test.y1, colors[1])
		Triangle(&unfilledDisplay, test.x1, test.y1, test.x0, test.y0, test.x2, test.y2, colors[2])
		Triangle(&unfilledDisplay, test.x1, test.y1, test.x2, test.y2, test.x0, test.y0, colors[3])
		Triangle(&unfilledDisplay, test.x2, test.y2, test.x0, test.y0, test.x1, test.y1, colors[4])
		Triangle(&unfilledDisplay, test.x2, test.y2, test.x1, test.y1, test.x0, test.y0, colors[5])

		err := compareWithReference(fmt.Sprintf("TestTriangles_%v", test.name), unfilledDisplay)
		if err != nil {
			t.Errorf("Fail %v", err)
		}

		filledDisplay := newTestDisplay()

		//Test all permutations of coordinates
		FilledTriangle(&filledDisplay, test.x0, test.y0, test.x1, test.y1, test.x2, test.y2, colors[0])
		FilledTriangle(&filledDisplay, test.x0, test.y0, test.x2, test.y2, test.x1, test.y1, colors[1])
		FilledTriangle(&filledDisplay, test.x1, test.y1, test.x0, test.y0, test.x2, test.y2, colors[2])
		FilledTriangle(&filledDisplay, test.x1, test.y1, test.x2, test.y2, test.x0, test.y0, colors[3])
		FilledTriangle(&filledDisplay, test.x2, test.y2, test.x0, test.y0, test.x1, test.y1, colors[4])
		FilledTriangle(&filledDisplay, test.x2, test.y2, test.x1, test.y1, test.x0, test.y0, colors[5])

		err = compareWithReference(fmt.Sprintf("TestTriangles_%v_Filled", test.name), filledDisplay)
		if err != nil {
			t.Errorf("Fail %v", err)
		}
	}
}


tinydraw.go
package tinydraw // import "tinygo.org/x/tinydraw"

import (
	"image/color"

	"errors"

	"tinygo.org/x/drivers"
)

// Line draws a line between two points
func Line(display drivers.Displayer, x0 int16, y0 int16, x1 int16, y1 int16, color color.RGBA) {
	if x0 == x1 {
		if y0 > y1 {
			y0, y1 = y1, y0
		}
		for ; y0 <= y1; y0++ {
			display.SetPixel(x0, y0, color)
		}
	} else if y0 == y1 {
		if x0 > x1 {
			x0, x1 = x1, x0
		}
		for ; x0 <= x1; x0++ {
			display.SetPixel(x0, y0, color)
		}
	} else { // Bresenham
		dx := x1 - x0
		if dx < 0 {
			dx = -dx
		}
		dy := y1 - y0
		if dy < 0 {
			dy = -dy
		}
		steep := dy > dx
		if steep {
			x0, x1, y0, y1 = y0, y1, x0, x1
		}
		if x0 > x1 {
			x0, x1, y0, y1 = x1, x0, y1, y0
		}
		dx = x1 - x0
		dy = y1 - y0
		ystep := int16(1)
		if dy < 0 {
			dy = -dy
			ystep = -1
		}
		err := dx / 2
		for ; x0 <= x1; x0++ {
			if steep {
				display.SetPixel(y0, x0, color)
			} else {
				display.SetPixel(x0, y0, color)
			}
			err -= dy
			if err < 0 {
				y0 += ystep
				err += dx
			}
		}
	}
}

// Rectangle draws a rectangle given a point, width and height
func Rectangle(display drivers.Displayer, x int16, y int16, w int16, h int16, color color.RGBA) error {
	if w <= 0 || h <= 0 {
		return errors.New("empty rectangle")
	}
	Line(display, x, y, x+w-1, y, color)
	Line(display, x, y, x, y+h-1, color)
	Line(display, x+w-1, y, x+w-1, y+h-1, color)
	Line(display, x, y+h-1, x+w-1, y+h-1, color)
	return nil
}

// FilledRectangle draws a filled rectangle given a point, width and height
func FilledRectangle(display drivers.Displayer, x int16, y int16, w int16, h int16, color color.RGBA) error {
	if w <= 0 || h <= 0 {
		return errors.New("empty rectangle")
	}
	for i := x; i < x+w; i++ {
		Line(display, i, y, i, y+h-1, color)
	}
	return nil
}

// Circle draws a circle given a point and radius
func Circle(display drivers.Displayer, x0 int16, y0 int16, r int16, color color.RGBA) {
	f := 1 - r
	ddfx := int16(1)
	ddfy := -2 * r
	x := int16(0)
	y := r
	display.SetPixel(x0, y0+r, color)
	display.SetPixel(x0, y0-r, color)
	display.SetPixel(x0+r, y0, color)
	display.SetPixel(x0-r, y0, color)
	for x < y {
		if f >= 0 {
			y--
			ddfy += 2
			f += ddfy
		}
		x++
		ddfx += 2
		f += ddfx

		display.SetPixel(x0+x, y0+y, color)
		display.SetPixel(x0-x, y0+y, color)
		display.SetPixel(x0+x, y0-y, color)
		display.SetPixel(x0-x, y0-y, color)
		display.SetPixel(x0+y, y0+x, color)
		display.SetPixel(x0-y, y0+x, color)
		display.SetPixel(x0+y, y0-x, color)
		display.SetPixel(x0-y, y0-x, color)
	}
}

// FilledCircle draws a filled circle given a point and radius
func FilledCircle(display drivers.Displayer, x0 int16, y0 int16, r int16, color color.RGBA) {
	f := 1 - r
	ddfx := int16(1)
	ddfy := -2 * r
	x := int16(0)
	y := r
	Line(display, x0, y0-r, x0, y0+r, color)
	for x < y {
		if f >= 0 {
			y--
			ddfy += 2
			f += ddfy
		}
		x++
		ddfx += 2
		f += ddfx

		Line(display, x0+x, y0-y, x0+x, y0+y, color)
		Line(display, x0+y, y0-x, x0+y, y0+x, color)
		Line(display, x0-x, y0-y, x0-x, y0+y, color)
		Line(display, x0-y, y0-x, x0-y, y0+x, color)
	}
}

// Triangle draws a triangle given three points
func Triangle(display drivers.Displayer, x0 int16, y0 int16, x1 int16, y1 int16, x2 int16, y2 int16, color color.RGBA) {
	Line(display, x0, y0, x1, y1, color)
	Line(display, x0, y0, x2, y2, color)
	Line(display, x1, y1, x2, y2, color)
}

// FilledTriangle draws a filled triangle given three points
func FilledTriangle(display drivers.Displayer, x0 int16, y0 int16, x1 int16, y1 int16, x2 int16, y2 int16, color color.RGBA) {
	if y0 > y1 {
		x0, y0, x1, y1 = x1, y1, x0, y0
	}
	if y1 > y2 {
		x1, y1, x2, y2 = x2, y2, x1, y1
	}
	if y0 > y1 {
		x0, y0, x1, y1 = x1, y1, x0, y0
	}

	if y0 == y2 { // y0 = y1 = y2 : it's a line
		a := x0
		b := x0
		if x1 < a {
			a = x1
		} else if x1 > b {
			b = x1
		}
		if x2 < a {
			a = x2
		} else if x2 > b {
			b = x2
		}
		Line(display, a, y0, b, y0, color)
		return
	}

	dx01 := x1 - x0
	dy01 := y1 - y0
	dx02 := x2 - x0
	dy02 := y2 - y0
	dx12 := x2 - x1
	dy12 := y2 - y1

	sa := int16(0)
	sb := int16(0)
	a := int16(0)
	b := int16(0)

	last := y1 - 1
	if y1 == y2 {
		last = y1
	}

	y := y0
	for ; y <= last; y++ {
		a = x0 + sa/dy01
		b = x0 + sb/dy02
		sa += dx01
		sb += dx02
		Line(display, a, y, b, y, color)
	}

	sa = dx12 * (y - y1)
	sb = dx02 * (y - y0)

	for ; y <= y2; y++ {
		a = x1 + sa/dy12
		b = x0 + sb/dy02
		sa += dx12
		sb += dx02
		Line(display, a, y, b, y, color)
	}
}


primitive.go
package primitive

import (
	"image/color"

	"github.com/tinygo-org/drivers"
)

// DrawLine draws a line between two points
func DrawLine(display drivers.Displayer, x0 int16, y0 int16, x1 int16, y1 int16, color color.RGBA) {
	if x0 == x1 {
		if y0 > y1 {
			y0, y1 = y1, y0
		}
		for ; y0 <= y1; y0++ {
			display.SetPixel(x0, y0, color)
		}
	} else if y0 == y1 {
		if x0 > x1 {
			x0, x1 = x1, x0
		}
		for ; x0 <= x1; x0++ {
			display.SetPixel(x0, y0, color)
		}
	} else { // Bresenham
		dx := x1 - x0
		if dx < 0 {
			dx = -dx
		}
		dy := y1 - y0
		if dy < 0 {
			dy = -dy
		}
		steep := dy > dx
		if steep {
			x0, x1, y0, y1 = y0, y1, x0, x1
		}
		if x0 > x1 {
			x0, x1, y0, y1 = x1, x0, y1, y0
		}
		dx = x1 - x0
		dy = y1 - y0
		ystep := int16(1)
		if dy < 0 {
			dy = -dy
			ystep = -1
		}
		err := dx / 2
		for ; x0 <= x1; x0++ {
			if steep {
				display.SetPixel(y0, x0, color)
			} else {
				display.SetPixel(x0, y0, color)
			}
			err -= dy
			if err < 0 {
				y0 += ystep
				err += dx
			}
		}
	}
}

// DrawRectangle draws a rectangle given a point and size
func DrawRectangle(display drivers.Displayer, x int16, y int16, w int16, h int16, color color.RGBA) {
	DrawLine(display, x, y, x+w, y, color)
	DrawLine(display, x, y, x, y+h, color)
	DrawLine(display, x+w, y, x+w, y+h, color)
	DrawLine(display, x, y+h, x+w, y+h, color)
}

// DrawFilledRectangle draws a filled rectangle given a point and size
func DrawFilledRectangle(display drivers.Displayer, x int16, y int16, w int16, h int16, color color.RGBA) {
	for i := x; i <= x+w; i++ {
		DrawLine(display, i, y, i, y+h, color)
	}
}

// DrawCircle draws a circle given a point and radius
func DrawCircle(display drivers.Displayer, x0 int16, y0 int16, r int16, color color.RGBA) {
	f := 1 - r
	ddfx := int16(1)
	ddfy := -2 * r
	x := int16(0)
	y := r
	display.SetPixel(x0, y0+r, color)
	display.SetPixel(x0, y0-r, color)
	display.SetPixel(x0+r, y0, color)
	display.SetPixel(x0-r, y0, color)
	for x < y {
		if f >= 0 {
			y--
			ddfy += 2
			f += ddfy
		}
		x++
		ddfx += 2
		f += ddfx

		display.SetPixel(x0+x, y0+y, color)
		display.SetPixel(x0-x, y0+y, color)
		display.SetPixel(x0+x, y0-y, color)
		display.SetPixel(x0-x, y0-y, color)
		display.SetPixel(x0+y, y0+x, color)
		display.SetPixel(x0-y, y0+x, color)
		display.SetPixel(x0+y, y0-x, color)
		display.SetPixel(x0-y, y0-x, color)
	}
}

// DrawFilledCircle draws a circle given a point and radius
func DrawFilledCircle(display drivers.Displayer, x0 int16, y0 int16, r int16, color color.RGBA) {
	f := 1 - r
	ddfx := int16(1)
	ddfy := -2 * r
	x := int16(0)
	y := r
	DrawLine(display, x0, y0-r, x0, y0+r, color)
	for x < y {
		if f >= 0 {
			y--
			ddfy += 2
			f += ddfy
		}
		x++
		ddfx += 2
		f += ddfx

		DrawLine(display, x0+x, y0-y, x0+x, y0+y, color)
		DrawLine(display, x0+y, y0-x, x0+y, y0+x, color)
		DrawLine(display, x0-x, y0-y, x0-x, y0+y, color)
		DrawLine(display, x0-y, y0-x, x0-y, y0+x, color)
	}
}

// DrawTriangle draws a triangle given three points
func DrawTriangle(display drivers.Displayer, x0 int16, y0 int16, x1 int16, y1 int16, x2 int16, y2 int16, color color.RGBA) {
	DrawLine(display, x0, y0, x1, y1, color)
	DrawLine(display, x0, y0, x2, y2, color)
	DrawLine(display, x1, y1, x2, y2, color)
}


// DrawFilledTriangle draws a filled triangle given three points
func DrawFilledTriangle(display drivers.Displayer, x0 int16, y0 int16, x1 int16, y1 int16, x2 int16, y2 int16, color color.RGBA) {
	if y0 > y1 {
		x0, y0, x1, y1 = x1, y1, x0, y0
	}
	if y1 > y2 {
		x1, y1, x2, y2 = x2, y2, x1, y1
	}
	if y0 > y1 {
		x0, y0, x1, y1 = x1, y1, x0, y0
	}

	if y0 == y2 { // y0 = y1 = y2 : it's a line
		a := x0
		b := x0
		if x1 < a {
			a = x1
		} else if x1 > b {
			b = x1
		}
		if x2 < a {
			a = x2
		} else if x2 > b {
			b = x2
		}
		DrawLine(display, a, y0, b, y0, color)
		return
	}

	dx01 := x1 - x0
	dy01 := y1 - y0
	dx02 := x2 - x0
	dy02 := y2 - y0
	dx12 := x2 - x1
	dy12 := y2 - y1

	sa := int16(0)
	sb := int16(0)
	a := int16(0)
	b := int16(0)

	last := y1 - 1
	if y1 == y2 {
		last = y1
	}

	for y := y0; y <= last; y++ {
		a = x0 + sa/dy01
		b = x0 + sb/dy02
		sa += dx01
		sb += dx02
		DrawLine(display, a, y, b, y, color)
	}

	sa = dx12 * (last - y1)
	sb = dx02 * (last - y0)

	for y := last; y <= y2; y++ {
		a = x1 + sa/dy12
		b = x0 + sb/dy02
		sa += dx12
		sb += dx02
		DrawLine(display, a, y, b, y, color)
	}
}


examples/pybadge/main.go
package main

import (
	"image/color"
	"machine"

	"tinygo.org/x/tinydraw"

	"tinygo.org/x/drivers/st7735"
)

func main() {
	machine.SPI1.Configure(machine.SPIConfig{
		SCK:       machine.SPI1_SCK_PIN,
		SDO:       machine.SPI1_SDO_PIN,
		SDI:       machine.SPI1_SDI_PIN,
		Frequency: 8000000,
	})

	display := st7735.New(machine.SPI1, machine.TFT_RST, machine.TFT_DC, machine.TFT_CS, machine.TFT_LITE)
	display.Configure(st7735.Config{
		Rotation: st7735.ROTATION_90,
	})

	display.FillScreen(color.RGBA{255, 255, 255, 255})

	tinydraw.Line(&display, 10, 10, 94, 10, getRainbowRGB(0))
	tinydraw.Line(&display, 94, 16, 10, 16, getRainbowRGB(15))
	tinydraw.Line(&display, 10, 20, 10, 118, getRainbowRGB(30))
	tinydraw.Line(&display, 16, 118, 16, 20, getRainbowRGB(45))

	tinydraw.Line(&display, 40, 40, 80, 80, getRainbowRGB(60))
	tinydraw.Line(&display, 40, 40, 80, 70, getRainbowRGB(75))
	tinydraw.Line(&display, 40, 40, 80, 60, getRainbowRGB(90))
	tinydraw.Line(&display, 40, 40, 80, 50, getRainbowRGB(105))
	tinydraw.Line(&display, 40, 40, 80, 40, getRainbowRGB(120))

	tinydraw.Line(&display, 100, 100, 40, 100, getRainbowRGB(135))
	tinydraw.Line(&display, 100, 100, 40, 90, getRainbowRGB(150))
	tinydraw.Line(&display, 100, 100, 40, 80, getRainbowRGB(165))
	tinydraw.Line(&display, 100, 100, 40, 70, getRainbowRGB(180))
	tinydraw.Line(&display, 100, 100, 40, 60, getRainbowRGB(195))
	tinydraw.Line(&display, 100, 100, 40, 50, getRainbowRGB(210))

	tinydraw.Rectangle(&display, 30, 106, 120, 20, getRainbowRGB(225))
	tinydraw.FilledRectangle(&display, 34, 110, 112, 12, getRainbowRGB(130))

	tinydraw.Circle(&display, 120, 30, 20, getRainbowRGB(240))
	tinydraw.FilledCircle(&display, 120, 30, 16, getRainbowRGB(145))

	tinydraw.Triangle(&display, 120, 102, 100, 80, 152, 46, getRainbowRGB(155))
	tinydraw.FilledTriangle(&display, 120, 98, 104, 80, 144, 54, getRainbowRGB(90))

	display.Display()

}

func getRainbowRGB(i uint8) color.RGBA {
	if i < 85 {
		return color.RGBA{i * 3, 255 - i*3, 0, 255}
	} else if i < 170 {
		i -= 85
		return color.RGBA{255 - i*3, 0, i * 3, 255}
	}
	i -= 170
	return color.RGBA{0, i * 3, 255 - i*3, 255}
}
